client/server-connection.cpp

```diff
@@ -13,159 +13,266 @@ const std::string_view LOCAL_HOST = "127.0.0.1";
-		msg.encode_header();
-		Send(msg);
+		MessageOut msg;
+		msg.FromString(message);
+		Send(std::move(msg));
 	}
 }
 
-void ServerConnection::Send(ServerMessage& msg) {
+void ServerConnection::Send(MessagePool::ptr_type msg) {
+	bool write_in_progress;
 	try {
-		asio::write(this->socket, asio::buffer(msg.GetDataPTR(), msg.length()));
+		std::lock_guard<std::mutex> guard(write_msg_mutex);
+		write_in_progress = !write_msg_queue.empty();
+		write_msg_queue.push_back(msg);
 	}
-	catch (std::exception e) {
-		std::cerr << "ServerConnection::Send(ServerMessage&): asio::write:" << e.what() << std::endl;
+	catch (std::exception& e) {
+		_log->error("ServerConnection::Send(): {}", e.what());
+		Disconnect();
+		return;
+	}
+	if (!write_in_progress) {
+		do_write();
 	}
 }
+void ServerConnection::Send(MessageOut& msg) {
+	bool write_in_progress;
+	try {
+		std::lock_guard<std::mutex> guard(write_msg_mutex);
+		write_in_progress = !write_msg_queue.empty();
+		for (auto& msg_ptr : msg.GetMessages()) {
+			write_msg_queue.push_back(msg_ptr);
+		}
+	}
+	catch (std::exception& e) {
+		_log->error("ServerConnection::Send(): {}", e.what());
+		Disconnect();
+		return;
+	}
+	if (!write_in_progress) {
+		do_write();
+	}
+}
+void ServerConnection::Send(MessageOut&& msg) { Send(msg); }
 
 void ServerConnection::RegisterConnectFunc(std::function<void()> func) { this->onConnect = std::move(func); }
 
 void ServerConnection::read_body() {
-	asio::error_code error = asio::error::eof;
-	asio::read(this->socket, asio::buffer(current_read_msg.GetBodyPTR(), current_read_msg.GetBodyLength()), error);
+	asio::async_read(
+			this->socket,
+			this->current_read_msg->buffer_body(),
+			[this](const asio::error_code& error, std::size_t /*length*/) {
+				if (error) {
+					this->socket.close();
+					throw asio::system_error(error, "read_body");
+				}
+				auto _log = spdlog::get("console_log");
 
-	if (!error) {
-		for (auto handler : this->message_handlers[current_read_msg.GetMessageType()]) {
-			handler(current_read_msg);
-		}
-	}
-	else if (error) {
-		this->socket.close();
-		throw asio::system_error(error);
-	}
+				MessagePool::ptr_type last_read_msg = current_read_msg;
+				current_read_msg = MessagePool::get();
+				uint32_t current_msg_id = last_read_msg->GetMessageID();
+				auto message_iter = read_messages.find(current_msg_id);
+				auto last_type = last_read_msg->GetMessageType();
+
+				if (last_type == SYNC) {
+					this->SyncHandler(last_read_msg.get());
+				}
+				else if (last_type == MessageType::MULTI_PART) {
+					if (message_iter == read_messages.cend()) {
+						auto message_in = std::make_unique<MessageIn>();
+						message_in->PushMessage(last_read_msg);
+						this->read_messages[current_msg_id] = std::move(message_in);
+					}
+					else {
+						message_iter->second->PushMessage(last_read_msg);
+					}
+				}
+				else {
+					// used to hold onto a single use MessageIn object for this scope
+					std::unique_ptr<MessageIn> message_in_unique;
```

Comment:
Could this just be a stack variable then?

Reply:
after getting some sleep, yes, I see what you mean here.